src/tree.ml

```diff
@@ -0,0 +1,711 @@
+        non linearity constraints. *)
+    ; nl_conds : PSet.t
+    (** Set of convertibility constraints [(i,j)] with [i < j]. The constraint
+        [(i,j)] is satisfied if the terms stored at indices [i] and [j] in the
+        [vars] array of the {!val:Eval.tree_walk} function are convertible. *)
+    ; fv_conds : (tvar array) IntMap.t
+    (** A mapping of [i] to [xs] represents a free variable condition that can
+        only be satisfied if only the free variables of [x] appear in the term
+        stored at slot [i] in the [vars] array of {!val:Eval.tree_walk}. *) }
+
+  (** [empty] is the condition pool holding no condition. *)
+  let empty : t =
+    { variables = IntMap.empty
+    ; nl_conds = PSet.empty
+    ; fv_conds = IntMap.empty }
+
+  (** [pp oc pool] prints condition pool [pool] to channel [oc]. *)
+  let pp : t pp = fun oc pool ->
+    let pp_fv oc fv_conds =
+      let pp_sep oc () = Format.pp_print_string oc ";" in
+      let pp_tvs = Format.pp_print_list ~pp_sep Print.pp_tvar in
+      let ppit oc (a, b) =
+        Format.fprintf oc "@[(%a@@%d)@]" pp_tvs (Array.to_list b) a
+      in
+      let pp_sep oc () = Format.pp_print_string oc "::" in
+      Format.fprintf oc "@[<v>@[%a@]@]"
+        (Format.pp_print_list ~pp_sep ppit) (IntMap.bindings fv_conds)
+    in
+    let pp_partials oc partials =
+      let pp_sep oc _ = Format.pp_print_string oc ";" in
+      let pp_part oc (i, j) = Format.fprintf oc "@[(%d↦%d)@]" i j in
+      Format.fprintf oc "@[<h>%a@]" (Format.pp_print_list ~pp_sep pp_part)
+        (IntMap.bindings partials)
+    in
+    let pp_available oc available =
+      let pp_av oc (i, j) = Format.fprintf oc "@[%d≡%d@]" i j in
+      let pp_sep = Format.pp_print_cut in
+      Format.fprintf oc "@[<hov>%a@]" (Format.pp_print_list ~pp_sep pp_av)
+        (PSet.elements available)
+    in
+    Format.fprintf oc "@[<hov>%a/@,%a/@,%a@]" pp_fv pool.fv_conds
+      pp_partials pool.variables pp_available pool.nl_conds
+
+  (** [is_empty pool] tells whether the pool of constraints is empty. *)
+  let is_empty : t -> bool = fun pool ->
+    PSet.is_empty pool.nl_conds && IntMap.is_empty pool.fv_conds
+
+  (** [register_nl slot i pool] registers the fact that the slot [slot] in the
+      [vars] array correspond to a term stored at index [i] in the environment
+      for the RHS. The first time that such a slot is associated to [i], it is
+      registered to serve as a reference point for testing convertibility when
+      (and if) another such slot is (ever) encountered. When that is the case,
+      a convertibility constraint is registered between the term stored in the
+      slot [slot] and the term stored in the reference slot. *)
+  let register_nl : int -> int -> t -> t = fun slot i pool ->
+    try
+      (* We obtain the reference slot (if any). *)
+      let r_slot = IntMap.find i pool.variables in
+      let cond = if r_slot < slot then (r_slot, slot) else (slot, r_slot) in
```

Comment:
Is it possible that `r_slot = slot` here? More generally, I'm wondering if we have the right representation for the convertibility constraints. I think we should merge all constraints tied to a particular environment slot. This means that we would map environment indices to sets of slots in the `vars` array. This way we would avoid potential normalization issues. Are constraints always checked after all filtering has been done on constructors?